libcontainer/dmz/dmz_linux.go

```diff
@@ -6,21 +6,32 @@ package dmz
 import (
 	"bytes"
 	"debug/elf"
-	_ "embed"
+	"embed"
 	"os"
 
 	"github.com/sirupsen/logrus"
 )
 
-// Try to build the runc-dmz binary. If it fails, replace it with an empty file
-// (this will trigger us to fall back to a clone of /proc/self/exe). Yeah, this
-// is a bit ugly but it makes sure that weird cross-compilation setups don't
-// break because of runc-dmz.
+// Try to build the runc-dmz binary on "go generate". If it fails (or
+// libcontainer is being imported as a library), the embed.FS will not contain
+// the file, which will then cause us to fall back to a clone of
+// /proc/self/exe.
 //
-//go:generate sh -c "make -B runc-dmz || echo -n >runc-dmz"
-//go:embed runc-dmz
+// There is an empty file called runc-dmz-dummy-file.txt in libcontainer/dmz in
+// order to work around the restriction that go:embed requires at least one
+// file to match the pattern.
+//
+//go:generate sh -c "make -B runc-dmz"
+//go:embed runc-dmz*
+var runcDmzFs embed.FS
+
+// The contents of runc-dmz.
 var runcDmzBinary []byte
 
+func init() {
+	runcDmzBinary, _ = runcDmzFs.ReadFile("runc-dmz")
```

Comment:
I'm thinking, since we only need dmz binary for runc start/run/exec, maybe instead of adding `init()` it is better to integrate this code into the `Binary` function below?

We may even get rid of using `bytes.NewBuffer` since it might be sufficient to do `runDmzFs.Open` and return the resulting file.

Reply:
Because we need to check the ELF prefix, having a single slice that is re-used for every container start is IMHO preferable to using the `fs.File` abstraction (where `Read` can return errors even though in practice it will never error).

I can switch to `sync.Once` if you like.